packages/core/src/tweening/spring.ts

```diff
@@ -0,0 +1,200 @@
+    if (typeof settleToleranceOrOnProgress === 'function') {
+      settleToleranceOrOnProgress(value, time);
+    } else if (typeof onProgressOrOnEnd === 'function') {
+      onProgressOrOnEnd(value, time);
+    }
+  };
+
+  spring = spring ?? {
+    mass: 0.05,
+    stiffness: 10,
+    damping: 0.5,
+  };
+
+  if (spring.mass <= 0) {
+    useLogger().error(new Error('Spring mass must be greater than 0.'));
+    return;
+  }
+  if (spring.stiffness < 0) {
+    useLogger().error(
+      new Error('Spring stiffness must be greater or equal to 0.'),
+    );
+    return;
+  }
+  if (spring.damping < 0) {
+    useLogger().error(
+      new Error('Spring damping must be greater or equal to 0.'),
+    );
+    return;
+  }
+
+  const project = useProject();
+  const thread = useThread();
+
+  const startTime = thread.time();
+
+  // figure out the step length for each frame
+
+  let position = from;
+  let velocity = spring.initialVelocity ?? 0;
+  let settled = false;
+
+  const update = (dt: number) => {
+    if (spring === null) {
+      return;
+    }
+    const positionDelta = position - to;
+
+    // Using hooks law: F=-kx; with k being the spring constant and x the offset
+    // to the settling position
+    const force = -spring.stiffness * positionDelta - spring.damping * velocity;
+
+    //updating the velocity
+    velocity += (force / spring.mass) * dt;
+
+    // apply velocity to the position
+    position += velocity * dt;
+  };
+
+  // Set simulation constant framerate
+  const simulationFrames = 120;
+
+  // Calculate a timestep based on on the simulation framerate
+  const timeStep = 1 / simulationFrames;
+
+  // Calculate the number steps needed for each frame of the project
+  const stepsPerFrame = simulationFrames / project.framerate;
+
+  let lastUpdate = startTime;
+
+  onProgress(from, 0);
+
+  // That is a stash we push leftover timesteps to
+  let accumulation = 0;
+
+  while (!settled) {
+    // We wait until the thread starts to account for offsets like
+    // `waitFor(0.005)`
+
+    const timeOffset = project.time - lastUpdate;
+
+    if (timeOffset <= 0) {
+      yield;
+      continue;
+    }
+    lastUpdate = project.time;
+
+    // Add the timeSteps for the current frame to the stash
+    accumulation += stepsPerFrame * timeStep;
+
+    let divider = 1;
+
+    // For the case we have uneven `timeStep` sizes, like when we have a
+    // framerate of 25 and a simulation framerate of 60 we end up having
+    // 2.4 steps per frame. So we just make substeps until the stash
+    // has less then 0.001 steps left.
+    while (divider < 10000) {
+      // Calculate the substep size based on the divider
+      const step = timeStep / divider;
+
+      // Sub step until there are no longer sub steps available
+      while (accumulation / step >= 1) {
+        update(step);
+        // remove the substeps from the stash
+        accumulation -= step;
+      }
+
+      // Increase the divider
+      divider *= 10;
+    }
```

Comment:
I think we can simplify that by calling `update` once, for the missing fraction of a frame.
For example, let's assume that a frame takes 33ms (30FPS) while the physics should be integrated using 10ms intervals (100FPS), then we should perform four updates per frame:
```
update(10ms);
update(10ms);
update(10ms);
update(3ms); // missing fraction
// 10 + 10 + 10 + 3 = 33ms
```
An update loop for something like this could look as follows:
```ts
const startTime = thread.time();
let simulationTime = startTime;

while (!settled) {
  // while our simulation is behind the project time
  while (simulationTime < project.time) {
    // calculate the remaining difference:
    const difference = project.time - simulationTime;

    // if the remaining time is smaller than our simulation step
    if (timeStep > difference) {
      // update using the remaining difference
      update(difference);
      // make the simulation time equal to the project time to make sure the while loop ends.
      simulationTime = project.time;
    } else {
      // otherwise update using the simulation step
      update(timeStep);
      // progress the simulation time by the same amount
      simulationTime += timeStep;
    }
  }
  
  // handle settling and call onProgress
}
```
Because we initialize the `simulationTime` using the thread time, it will also account for the thread offset.

Reply:
I tried your approach but I don't really know whats wrong with it.

https://user-images.githubusercontent.com/35671734/219866199-a083c89c-5c3a-4e28-9b8b-d02bd1f7b790.mov

Reply:
Oh yeah, forgot about this, you also need to change this:
```ts
Math.abs(to) - Math.abs(position) < settleTolerance
```
to
```ts
Math.abs(to - position) < settleTolerance
```

Otherwise, a spring starting at -400 with the goal of 400 will end immediately, because the check would succeed

Reply:
Okay that works. But now the animation is not that constant in duration anymore across different framerates. 
But yeah, I assume that smooth animations is more important than duration consistency.

Reply:
You're right, I absolutely forgot about the duration.
This would require some modifications to the loop:

```ts
const startTime = thread.time();
let simulationTime = startTime;

while (!settled) {
  while (simulationTime < project.time) {
    const difference = project.time - simulationTime;

    if (timeStep > difference) {
      update(difference);
      simulationTime = project.time;
    } else {
      update(timeStep);
      simulationTime += timeStep;
    }

    // perform the check during every iteration:
    if (
      Math.abs(to - position) < settleTolerance &&
      Math.abs(velocity) < settleTolerance
    ) {
      // set the thread time to simulation time:
      thread.time(simulationTime);
      settled = true;
      // break out when settled
      break;
    }
  }

  // only yield if we haven't settled yet.
  if (!settled) {
    onProgress(position, project.time - startTime);
    yield;
  }
}
```
This will make the durations much more aligned between framerates.
The solution is not perfect, but the problem is really complex and so I think we can leave it like this for now.

Reply:
Okay, will apply your modifications. Thanks a lot! :-)